src/Features/Core/Portable/Completion/CompletionHelper.cs

```diff
@@ -0,0 +1,525 @@
+            {
+                return -1;
+            }
+            else if (item2.Rules.Preselect && !item1.Rules.Preselect)
+            {
+                return 1;
+            }
+
+            diff = PatternMatch.CompareCase(match1, match2);
+            if (diff != 0)
+            {
+                return diff;
+            }
+
+            diff = PatternMatch.ComparePunctuation(match1, match2);
+            if (diff != 0)
+            {
+                return diff;
+            }
+
+            return 0;
+        }
+
+        private bool IsArgumentName(CompletionItem item)
+        {
+            return item.Tags.Contains(CompletionTags.ArgumentName);
+        }
+
+        /// <summary>
+        /// Returns true if the character is one that can commit the specified completion item. A
+        /// character will be checked to see if it should filter an item.  If not, it will be checked
+        /// to see if it should commit that item.  If it does neither, then completion will be
+        /// dismissed.
+        /// </summary>
+        public bool IsCommitCharacter(CompletionItem item, char ch, string textTypedSoFar, string textTypedWithChar = null)
+        {
+            // general rule: if the filtering text exactly matches the start of the item then it must be a filter character
+            textTypedWithChar = textTypedWithChar ?? textTypedSoFar + ch;
+            if (item.DisplayText.StartsWith(textTypedWithChar, StringComparison.CurrentCultureIgnoreCase)
+                || item.FilterText.StartsWith(textTypedWithChar, StringComparison.CurrentCultureIgnoreCase))
+            {
+                return false;
+            }
+
+            foreach (var rule in item.Rules.CommitCharacterRules)
+            {
+                switch (rule.Kind)
+                {
+                    case CharacterSetModificationKind.Add:
+                        if (rule.Characters.IndexOf(ch) >= 0)
+                            return true;
+                        break;
+
+                    case CharacterSetModificationKind.Remove:
+                        if (rule.Characters.IndexOf(ch) >= 0)
+                            return false;
+                        break;
+
+                    case CharacterSetModificationKind.Replace:
+                        return rule.Characters.IndexOf(ch) >= 0;
+                }
+            }
+
+            return _rules.DefaultCommitCharacters.IndexOf(ch) >= 0;
+        }
+
+        /// <summary>
+        /// Returns true if the character typed should be used to filter the specified completion
+        /// item.  A character will be checked to see if it should filter an item.  If not, it will be
+        /// checked to see if it should commit that item.  If it does neither, then completion will
+        /// be dismissed.
+        /// </summary>
+        public bool IsFilterCharacter(CompletionItem item, char ch, string textTypedSoFar, string textTypedWithChar = null)
```

Comment:
This logic is nearly identical to IsCommitCharacter.  It seems like we'd be able to share most of this in a helper.